dali/operators/reader/loader/video/video_loader_decoder_gpu.cc

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/video/video_loader_decoder_gpu.h"
+
+#include "dali/util/nvml.h"
+#include "dali/core/cuda_stream.h"
+
+namespace dali {
+void VideoSampleGpu::Decode() {
+  TensorShape<4> shape = {
+    sequence_len_,
+    video_file_->Height(),
+    video_file_->Width(),
+    video_file_->Channels()};
+
+  data_.Resize(
+    shape,
+    DALIDataType::DALI_UINT8);
+
+  for (int i = 0; i < sequence_len_; ++i) {
+    int frame_id = span_->start_ + i * span_->stride_;
+    video_file_->SeekFrame(frame_id);
+    video_file_->ReadNextFrame(
+      static_cast<uint8_t *>(data_.raw_mutable_data()) + i * video_file_->FrameSize());
+  }
+}
+
+VideoLoaderDecoderGpu::~VideoLoaderDecoderGpu() {
+  CUDA_CALL(cudaStreamDestroy(cuda_stream_));
+}
+
+cudaStream_t VideoLoaderDecoderGpu::GetCudaStream() {
+  #if NVML_ENABLED
+  {
+    nvml::Init();
+    static float driver_version = nvml::GetDriverVersion();
+    if (driver_version > 460 && driver_version < 470.21) {
+      DALI_WARN_ONCE("Warning: Decoding on a default stream. Performance may be affected.");
+      return 0;
+    }
+  }
+  #else
+  {
+    int driver_cuda_version = 0;
+    CUDA_CALL(cuDriverGetVersion(&driver_cuda_version));
+    if (driver_cuda_version >= 11030 && driver_cuda_version < 11040) {
+      DALI_WARN_ONCE("Warning: Decoding on a default stream. Performance may be affected.");
+      return 0;
+    }
+  }
+  #endif
+
+  // TODO(awolant): Check per decoder stream
+  cudaStream_t stream;
+  DeviceGuard dg(device_id_);
+  CUDA_CALL(cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking));
```

Comment:
Consider using `CUDAStream`, or even better, just lease one from the pool:
```
dali::CUDAStreamPool::instance().Get(device_id_);
```

Reply:
We can't use `CUDAStream` as this is derived from `UniqueHandle` and I want to share this stream between decoders for now.